include/maths/CBoostedTreeFactory.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_maths_CBoostedTreeFactory_h
+#define INCLUDED_ml_maths_CBoostedTreeFactory_h
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+#include <core/CPackedBitVector.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTree.h>
+
+#include <numeric>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace ml {
+namespace maths {
+
+class CNode;
+class CBoostedTreeImpl;
+
+//! SimpleFactory for CBoostedTree object
+class CBoostedTreeFactory final {
```

Comment:
I think the natural thing to do if we're introducing a factory and given `CBoostedTree` inherits from `CDataFrameRegressionModel` would be to have a base class for the factory type as well and build method which returns `TDataFrameRegressionModelUPtr`. Since we don't have other implementations at present this isn't strictly needed, but it would be another advantage of having the factory.

Reply:
I see your point of having abstract factory for symmetry of class hierarchies. However, I would  not introduce this level of abstraction prematuraly. Say, when there is another regression model we need to construct, we introduce polymorphism.

Reply:
Ok, I'm happy to keep as it is now: it makes the change smaller. (I suspect we will want to provide some additional models down-the-line. At this point if we want to allow the user to manually select the type, then injecting this choice by an appropriate factory object will be in the back pocket :).)